rfcs/037-index-query-composition.md

```diff
@@ -0,0 +1,115 @@
+- Feature Name: Index queries and query composition
+- Start Date: 2021-05-12
+- RFC PR: [amundsen-io/rfcs#0000](https://github.com/amundsen-io/rfcs/pull/0000) (after opening the RFC PR, update this with a link to it and update the file name)
+- Amundsen Issue: [amundsen-io/amundsen#0000](https://github.com/amundsen-io/amundsen/issues/0000) (leave this empty for now)
+
+# Index queries and query composition
+
+## Summary
+
+Add indexing of queries, and the components of the query, to Amundsen. The components of the query included in this RFC are where clauses and joins.
+
+## Motivation
+
+Queries are one of the most common ways that users interact with data and there is a significant amount of latent information that exists within the queries that have been executed which can be useful for users to discover. By indexing queries and their individual components Amundsen will be able to answer questions such as:
+
+- What tables are most commonly joined to this table and how?
+- Are there any filters (where clauses) that are always applied when accessing this table?
+- What real-world queries are being used on this table?
+- What queries are relevant now, given the recent queries that have been executed?
+
+This information helps users to discover _how to use_ their data.
+
+## Guide-level Explanation (aka Product Details)
+
+Four new asset types are added to Amundsen as depicted in the following diagram:
+
+![Query Assets](../assets/037/query-assets.png)
+
+1. **Query**: these contain the full SQL for a query as well as relationships to Tables, Users, Query Executions, Joins and Wheres
+2. **Query Execution**: this represents the total number of times a query was executed within a time window. By having query executions separated from queries Amundsen can keep one copy of the query in the database and sum the total number of executions at any point in time to get the total execution count. Query Executions can be added or dropped to add or remove relevance to a Query, Join or Where
+3. **Join**: this is a component of a query, a single join clause (e.g. `table_1 a join table_2 b on a.field = b.field`). Joins are connected to the columns they are joined on as well as to queries. By connecting to queries, Amundsen can rank order the most frequent joins by accessing the query executions.
+4. **Where**: this is a component of a query, a single where clause (e.g. `where field_a is not null`). Wheres are connected to columns as well as to queries. By connecting to queries, Amundsen can rank order the most frequent wheres by accessing the query executions.
```

Comment:
I am not sure actually, for join relationship it makes SENSE. For where , it seems to be a bit specific use case. Could you talk about this use case? And there are many different sql paradigm as well(e.g ctas), why only where clause?

Reply:
The use case we've seen is an analyst or data engineer who is trying to figure out if there are any common filter conditions. Usually that's to see if there are any caveats with the data (e.g. you basically always need to filter out NULLs from the column, or you need to exclude this status to make sure to only have prod data). Sometimes this is used to identify places to add new aggregate tables (e.g. if everyone is using this filter, might make sense to create a new derived table).

When you say CTAS, do you mean `create table as select?` If so, I think the way we'd like to ingest that is as lineage, which is out of scope for here

Reply:
@feng-tao could you please elaborate on `I am not sure actually, for join relationship it makes SENSE`. I'm not sure I completely follow the intent behind the question.

In regards to the where clause, the response from @dorianj is spot on but I will elaborate on the point about it being a specific use-case. The structure of this where-clause is actually set up such that it can capture relationships between any number of tables and columns in order to handle both complex as well as simple scenarios. This setup supports the following (and more):

- Simple:
```sql
where table1.field = 'value'
```
- Multiple columns
```sql
where table1.field > table2.another_value * 10
```
- Sub query
```sql
where table1.date = (select max(dt) from date_tbl)
```

Let me know if this helps.

Reply:
@feng-tao just wanted to bump this to see if it answers your questions or if you're able to add some additional context.

Reply:
I think adding a query model to show it in table optional makes sense. I am just not convinced showing `where` relationship is useful for end-user. But it is not a blocker from my side.

Reply:
or at the very least that needs to make it more generic like `query->transform->table` etc